### DISCUSSION TITLE
How to use `<>` for Seq

### DISCUSSION THREAD

**Discussion by oharboe** [chipsalliance/chisel#4336] [General]
Is there a way to use `<>` for Seq?

So I have a local variable Seq[Foo] of modules, and a val bar=Vec(n, new FooBundlke()) in my io and I want to hook the Foos to the io...

```
  foos.map(_.io.foo) <> io.foo
```

**Comment by oharboe** [comment]
@hovind FYI

**Comment by jackkoenig** [comment]
We should add this to Chisel itself (or better yet, implement `Connectable[Seq[T <: Data]]`, but it's pretty easy to add an extension method:
```scala
object Whatever {
  implicit class SeqExtensions[T <: Data](xs: Seq[T]) {
    def :<>=(rhs: Seq[T]): Unit =
      xs.zip(rhs).foreach { case (l, r) => l :<>= r }
  }
}
```

Then you just make sure to import `SeqExtensions` into your lexical scope:
```scala
import Whatever._
// or
import Whatever.SeqExtensions
```
Then you can use the extension operator:
```scala
foo.map(_.io.foo) :<>= io.foo
```

I am using `:<>=` instead of `<>` just because it has better behavior but you can of course do the same thing with `<>`.

**Reply by oharboe** [reply]
Is `<>` an anachronism?

**Reply by tymcauley** [reply]
Love the idea of adding this to Chisel :)

**Reply by jackkoenig** [reply]
> Is `<>` an anachronism?

I wouldn't quite say it's an anachronism but we are in a period of tactical ambiguity. `:<>=` has some nice advantages, most of which could be ported to `<>` over a deprecation period, however there are 2 that can't: 

1. Under Scala precedence rules, operators that end in `=` have lowest precedence, otherwise they get the precedence of the first character. `<>` has sort of bad precedence. If you try to use `<>` in conjunction with some other operators you get issues:
```scala
sum :<>= a + b // this works
sum <> a + b // this works because < has lower precedence than +
sum :<>= a & b // this works
sum <> a & b // value & is not a member of Unit (because < has higher precedence than &)
```
The annoying question is "does this matter?" People almost exclusively use `<>` and `:<>=` to bidirectionally connect Aggregates, it's very rare that you have operators on the right-hand side. The issue is just that when you do get bitten by this, the error message is so bad and confuses people quite a bit.

2. `:<>=` visually indicates that it is a combination of `:<=` and `:>=` which is true and a good thing to teach and encourage users to understand. `:<=` and `:>=` are uncommon but not unheard of (and a lot more common now that `.asTypeOf` returns a read-only value[^1]), and the relationship between `:<>=` and them is very clear.

On the negative side, `:<>=` is obviously a bit obnoxious to type and can turn new users off compared to the pithy `<>`. Not to mention that `<>` has always been the operator so all existing Chisel code is chock full of it.

So that leaves us with the question of what we should do? Some argue, and it would be by far the easiest for developers, is we should just deprecate `<>`. It will encourage existing users to learn and understand the benefits of `:<>=`. My preference is that we should "reclaim" `<>`. We should deprecate the behavior we don't like (lack of Scala type safety and implicit truncation), and then effectively make it an alias for `:<>=`.[^2]

If I can get people to agree with me on reclaiming `<>`, then there's just the issue of that deprecation process is a huge pain and no one (myself included) wants to do the work 😂. Thus... tactical ambiguity!

[^1]: For example, if you have a bidirectional "output" port that you want to give a default value of zero to, you used to be able to just do `myBidirSink :<>= 0.U.asTypeOf(myBidirSink)`. That worked because (as an implementation detail), `asTypeOf` returned a `Wire` so the flipped fields of `myBidirSink` would drive the flipped fields of that returned Wire. Now that `asTypeOf` returns a read-only reference, it's an error (warning in 6.x). The solution is to use `:<=`: `myBidirSink :<= 0.U.asTypeOf(myBidirSink)` which will only drive the aligned fields of `myBidirSink` and will not attempt to create the flipped connections.

[^2]: We could potentially solve (1) above with the compiler plugin, we could just hack the compiler to override the precedence and avoid that issue. Another possible solution is to change `<>` to return something other than `Unit`, then we could use macros (instead of the compiler plugin, macros are always easier to deal with) to add an implicit conversion from this new type back to the original type of the RHS (so that any functions called on it would "work") but in that implicit conversion trigger a compile-time error and tell the user about the precedence issue.

**Reply by oharboe** [reply]
Good to know!

Regardless of what happens, there are no plans to change the behavior of `:<>=`, so it is "safe" to use now?

Can I create a slide for our next planning meeting?

TL;DR Use `:<>=`

:-)

**Reply by jackkoenig** [reply]
`:<>=` is safe to use 🙂. Any changes would be API extensions, not deprecation or modifications.

**Reply by mwachs5** [reply]
> We should deprecate the behavior we don't like (lack of Scala type safety and implicit truncation),

The other major difference is the directionality inference of `<>`. Need to deprecate that one too.

**Reply by jackkoenig** [reply]
With more thought and discussion, I think we should just deprecate `<>` altogether.

**Reply by oharboe** [reply]
> With more thought and discussion, I think we should just deprecate `<>` altogether.

What tipped the scales?

Precedence rules?

**Comment by mwachs5** [comment]
I would encourage us if we add this to chisel to add a check that the sequences are the same length

**Comment by oharboe** [comment]
Example of how to do this without waiting for a Chisel upgrade: https://scastie.scala-lang.org/f8sVrCjcRuqOCARbgkx28w
